src/lib/shim/src/signals.rs

```diff
@@ -0,0 +1,363 @@
+    assert!(crate::global_allow_native_syscalls::get());
+    if sig == Signal::SIGKILL {
+        // No need to restore default action, and trying to do so would fail.
+    } else {
+        let action = sigaction::new_with_default_restorer(
+            SignalHandler::SigDfl,
+            SigActionFlags::empty(),
+            sigset_t::EMPTY,
+        );
+        unsafe { linux_api::signal::rt_sigaction(sig, &action, None) }.unwrap();
+    }
+    let pid = rustix::process::getpid();
+    rustix::process::kill_process(pid, rustix::process::Signal::from_raw(sig.into()).unwrap())
+        .unwrap();
+    unreachable!()
+}
+
+/// Handle pending unblocked signals, and return whether *all* corresponding
+/// signal actions had the SA_RESTART flag set.
+pub fn process_signals(mut ucontext: Option<&mut libc::ucontext_t>) -> bool {
+    let mut host = crate::global_host_shmem::get();
+    let mut process = crate::global_process_shmem::get();
+    let mut thread = crate::tls_thread_shmem::get();
+    let mut host_lock = host.protected().lock();
+
+    let mut restartable = true;
+
+    while let Some((sig, mut siginfo)) =
+        shim_shmem::take_pending_unblocked_signal(&host_lock, &process, &thread)
+    {
+        let blocked_signals = thread.protected.borrow(&host_lock.root).blocked_signals;
+        let action = *unsafe { process.protected.borrow(&host_lock.root).signal_action(sig) };
+
+        if matches!(unsafe { action.handler() }, SignalHandler::SigIgn) {
+            continue;
+        }
+
+        if matches!(unsafe { action.handler() }, SignalHandler::SigDfl) {
+            match defaultaction(sig) {
+                linux_api::signal::LinuxDefaultAction::IGN => continue,
+                linux_api::signal::LinuxDefaultAction::CORE
+                | linux_api::signal::LinuxDefaultAction::TERM => {
+                    drop(host_lock);
+                    die_with_fatal_signal(sig);
+                }
+                linux_api::signal::LinuxDefaultAction::STOP => unimplemented!(),
+                linux_api::signal::LinuxDefaultAction::CONT => unimplemented!(),
+            }
+        }
+
+        trace!("Handling emulated signal {sig:?}");
+
+        let handler_mask = {
+            let mut m = action.mask() | blocked_signals;
+            if !action.flags_retain().contains(SigActionFlags::SA_NODEFER) {
+                m.add(sig)
+            }
+            m
+        };
+        thread.protected.borrow_mut(&host_lock.root).blocked_signals = handler_mask;
+
+        if action.flags_retain().contains(SigActionFlags::SA_RESETHAND) {
+            unsafe {
+                *process
+                    .protected
+                    .borrow_mut(&host_lock.root)
+                    .signal_action_mut(sig) = sigaction::new_with_default_restorer(
+                    SignalHandler::SigDfl,
+                    SigActionFlags::empty(),
+                    sigset_t::EMPTY,
+                )
+            };
```

Comment:
leave a comment why this is safe (is `SigDfl`)?

Reply:
Done